### ISSUE TITLE
Add support for redirecting HTTP to HTTPS

### ISSUE THREAD

**Issue by JamesChenX** [reactor/reactor-netty#1511]
## Motivation

This should be a very common use case that a user sends an HTTP request to a Web server that uses TLS and the server better redirect HTTP to HTTPS rather than responding with an error message which is confusing for users.

## Desired solution

reactor-netty should use OptionalSslHandler instead of SslHandler so it's possible to redirect HTTP to HTTPS,
And the reference code to redirect HTTP to HTTPS is (Note that the code cannot work with SslHandler because SslHandler starts handshake processing at io.netty.handler.ssl.SslHandler#channelActive while we cannot detect whether the following request is HTTP or HTTPS):

```java
public class RedirectHttpsHandler extends ChannelInboundHandlerAdapter {

    private HttpRequestDecoder requestDecoder;
    private HttpResponseEncoder responseEncoder;
    private DelegatingChannelHandlerContext handlerContext;

    @Override
    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
        handlerContext = new DelegatingChannelHandlerContext(ctx) {
            @Override
            public ChannelHandlerContext fireChannelRead(Object msg) {
                if (msg instanceof HttpRequest) {
                    HttpMessage message = (HttpMessage) msg;
                    String address = message.headers().get(HOST);
                    if (address == null) {
                        ctx.close();
                        return ctx;
                    }
                    HttpResponse response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.PERMANENT_REDIRECT);
                    response.headers().set(LOCATION, "https://" + address);
                    try {
                        responseEncoder.write(ctx, response, new DefaultChannelPromise(ctx.channel()) {
                        });
                    } catch (Exception e) {
                        log.error(e);
                    }
                    ctx.channel().writeAndFlush(response)
                            .addListener(ChannelFutureListener.CLOSE);
                }
                return ctx;
            }
        };
        super.channelRegistered(ctx);
    }

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
        boolean encrypted = SslHandler.isEncrypted((ByteBuf) msg);
        if (!encrypted) {
            if (requestDecoder == null) {
                requestDecoder = new HttpRequestDecoder();
                responseEncoder = new HttpResponseEncoder();
            }
            // Redirect to another pipeline
            requestDecoder.channelRead(handlerContext, msg);
        } else {
            ctx.fireChannelRead(msg);
        }
    }

}
```

## Considered alternatives

No.

## Additional context

reactor-netty: 1.0.3

**Comment by violetagg** [comment]
@JamesChenX I suppose that this one can be used for this purpose https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/OptionalSslHandler.java

**Comment by JamesChenX** [comment]
@violetagg Yes, we need to use `OptionalSslHandler` instead of `SslHandler` (used by reactor.netty.tcp.TcpServer#secure currently) to implement redirecting HTTP to HTTPS.

I just think this is a very common feature and it's would be good news if reactor-netty can do this for developers and there are other developers who want this feature too.

**Comment by hc-codersatlas** [comment]
How would this work with. Reactor Netty only allowing a single port i.e. can't run 80 and 443 at the same time? You'd have to run 2 different instances so it wouldn't really matter if `SslHandler` is optional or not?

**Comment by JamesChenX** [comment]
@hc-codersatlas We don't need to listen to two ports and one HTTP server is enough because:
After a TCP established between the client and the server:
For the cases:
1. The client sends a TLS client hello to the server, then the server can just try to set up SSL and uses HTTPS.
2. The client sends an HTTP request to the server, the server can just respond with a status of 308 (redirect) to notify the client to redirect to HTTPS.
To aware of the two different cases, we should use `OptionalSslHandler` instead of `SslHandler`.

**Comment by violetagg** [comment]
@JamesChenX Are you interested in providing a PR? One interesting aspect of this feature would be: how Reactor Netty is expected to handle POST/PUT requests on behalf of the application i.e. before redirection what should be done with the incoming data.

**Comment by JamesChenX** [comment]
@violetagg OK, I can give a PR during 01/3~15/3 when I am available.
Here are some details:
1. Use 308 instead of 301 because:
>    The 308 (Permanent Redirect) status code indicates that the target
>    resource has been assigned a new permanent URI and any future
>    references to this resource ought to use one of the enclosed URIs.

According to https://tools.ietf.org/html/rfc7238 (The Hypertext Transfer Protocol Status Code 308 (Permanent Redirect))

2. It's reasonable for real applications to redirect any kind of HTTP request to HTTPS no matter its method, we redirect HTTP request of any kind of method (POST/PUT/GET...) to HTTPS by assigning a Location header in the response.

3. No need to support redirecting HTTP/2 because it's meaningless due to the facts
* Most web browsers don't support unencrypted HTTP2 at all.
* For encrypted HTTP2, the SSL session is already set up before we can know it's a HTTP2 (secure) request

**Comment by violetagg** [comment]
@JamesChenX thanks 👍
